nixos/rke2: merge code with nixos/k3s

maevii opened the pull request:
Resolves #445117, supersedes #401080 & #449578.

K3s and RKE2 share a lot of functionality, but since the nixpkgs RKE2 maintainer team is much smaller than K3s', its module tends to fall far behind - and for feature parity, code duplication across the two modules would be needed (see #401080). Merging them should make it possible for the RKE2 module to keep up with K3s without having maintainers manually copy code over.

I haven't removed or added any options, this PR is more or less just a (big) refactor. I also haven't tested the K3s module manually, but all tests in `nixos/tests/{k3s,rke2}` passed and I'm [currently using the RKE2 module](https://git.azey.net/infra/nixos-servers/src/commit/3a6740b61546c680632f0e0a6316b947abfa9b7d/config/rke2/default.nix#L15-L16) to host my site with no issues.

I've also added myself as a maintainer to RKE2.

Maintainer cc:
@rorosen @zimbatm @stefan-bordei
@NixOS/k3s 

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [x] [NixOS tests] in [nixos/tests].
  - [x] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [x] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

rorosen reviewed (commented):
This is great, thanks! Just some minor nits.

It would be nice to cover the auto deploy stuff in rke2 tests like in https://github.com/NixOS/nixpkgs/pull/401080

maevii committed:
nixos/k3s: factor out generic functionality

Also slightly reworded a few option descriptions, commit prepares for merge with nixos/rke2.

maevii committed:
rke2: add azey7f as maintainer

azey7f commented on nixos/modules/services/cluster/rancher/default.nix:
```diff
@@ -0,0 +1,906 @@
+            List of derivations that provide container images.
+            All images are linked to {file}`${imageDir}` before ${name} starts and consequently imported
+            by the ${name} agent. Consider importing the k3s airgap images archive of the k3s package in
+            use, if you want to pre-provision this node with all k3s container images. This option
```
Good catch, I missed this one. There are still a few references to the k3s docs left, but since the docs are nearly identical it should be fine IMO.

azey7f commented:
> It would be nice to cover the auto deploy stuff in rke2 tests like in https://github.com/NixOS/nixpkgs/pull/401080

@rorosen Really glad you mentioned this, apparently the `manifests`/`autoDeployCharts` options weren't working... RKE2 was trying to open the generated manifests as `O_RDWR`, which didn't work since they were just `/nix/store` links. I can't really think of any good way to fix this short of a patch or upstream PR, so I borrowed your solution of making the generated manifests JSON instead of YAML which seems to have worked.

In any case I copied your changes to the tests from the other PR, and after the JSON fix they ran perfectly out of the box. We should probably also be testing `autoDeployCharts`, but that would require `./../` linking to or duplicating the k3s test's helm chart, which feels a bit dirty so I didn't include anything for now.

azey7f commented:
@superherointj Fixed, the wiki seems to just point to the github so we can update it here. Couldn't find any other references to the module path by grepping through the repo, so this should be everything.

rorosen commented:
> RKE2 was trying to open the generated manifests as O_RDWR

That `O_RDWR` opening sucks, but templating JSON is fine as the auto deploy controller works with both, JSON and YAML. I also like how it's structured now in the sense that options and config can be extended and overridden. I think we can work very well with this structure in the future.

However, the `autoDeployCharts` option won't work for RKE2 as is. `autoDeployCharts` depends on static chart content that is placed in `/var/lib/rancher/k3s/server/static/` so the `%{KUBERNETES_API}%` special variable can be used in the `HelmChart` custom resource (as described in https://docs.k3s.io/helm#using-the-helm-controller). The placement of the static chart file is currently only used in the k3s module, so it won't work for rke2.

I wonder if it just works placing charts into `/var/lib/rancher/rke2/server/static/`, the RKE2 docs don't mention it but the rest looks similar enough to k3s that it may work the same way.

azey7f commented:
@rorosen You're right, I completely missed that `%{KUBERNETES_API}%` usage in `mkHelmChartCR`. I intentionally put `charts` into `k3s.nix` since RKE2 doesn't support `/server/static` from my testing.

According to https://github.com/rancher/rke2/issues/224, upstream packages their default charts by base64-encoding them into `chartValues`. The easiest way to implement from our side would probably be some sort of `runCommand`/`readFile` chain like what's used in `fromYaml`, I'll see what I can come up with.

maevii committed:
nixos/rke2: merge with nixos/k3s

K3s module was moved from `nixos/.../k3s` to `nixos/.../rancher`.

`rke2/default.nix` was moved to `rancher/rke2.nix`, and some options from RKE2 were migrated into the common `default.nix` for backwards compatibility.

Manifest generation was also changed, instead of multi-doc YAML files the module now generates `kind: List` manifests.

rorosen committed:
nixos/tests/rke2: use `manifests` and `images` options

Copied from #401080

rorosen commented on nixos/modules/services/cluster/rancher/default.nix:
```diff
@@ -0,0 +1,968 @@
+              "L+".argument = "${pkgs.writeText "config.toml.tmpl" cfg.containerdConfigTemplate}";
+            };
+          })
+          // (lib.mapAttrs' (n: v: mkChartRule n v) helmCharts);
```
```suggestion
          // (lib.optionalAttrs staticContentSupport (lib.mapAttrs' mkChartRule helmCharts));
```

rorosen reviewed (approved):
I still like this a lot. All k3s and rke2 tests pass locally with this patch applied. Good to go from my side
